packages/core/lib/command-utils.js

```diff
@@ -0,0 +1,209 @@
+    );
+  }
+
+  // determine whether Truffle is being run from the bundle or from ./cli.js
+  // and require commands accordingly
+  let command;
+  if (typeof BUNDLE_VERSION !== "undefined") {
+    const path = require("path");
+    const filePath = path.join(__dirname, `${chosenCommand}.bundled.js`);
+    // we need to use this library to bypass webpack's require which can't
+    // access the user's filesystem
+    const originalRequire = require("original-require");
+    command = originalRequire(filePath);
+  } else {
+    const filePath = `./commands/${chosenCommand}`;
+    command = require(filePath);
+  }
+
+  const yargs = require("yargs/yargs")();
+  yargs.command(require(`./commands/${chosenCommand}/meta`));
+  const commandOptions = yargs.parse(inputStrings);
+
+  // remove the task name itself put there by yargs
+  if (commandOptions._) commandOptions._.shift();
+
+  // several commands have a help property that is a function
+  if (typeof command.meta.help === "function") {
+    command.meta.help = command.meta.help(options);
+  }
+
+  return {
+    name: chosenCommand,
+    options: commandOptions,
+    run: command.run,
+    meta: command.meta
+  };
+};
+
+const prepareOptions = (command, inputStrings, options) => {
+  // some options might throw if options is a Config object
+  // if so, let's ignore those values
+  const clone = {};
+  Object.keys(options).forEach(key => {
+    try {
+      clone[key] = options[key];
+    } catch {
+      // do nothing with values that throw
+    }
+  });
+
+  // method `extractFlags(args)` : Extracts the `--option` flags from arguments
+  const inputOptions = extractFlags(inputStrings);
+
+  // adding allowed global options as enumerated in each command
+  const allowedGlobalOptions = command.meta.help.allowedGlobalOptions
+    .filter(tag => tag in globalCommandOptions)
+    .map(tag => globalCommandOptions[tag]);
+
+  const allValidOptions = [
+    ...command.meta.help.options,
+    ...allowedGlobalOptions
+  ];
+
+  const validOptions = allValidOptions.reduce((a, item) => {
+    // we split the options off from the arguments
+    // and then we split to handle options of the form --<something>|-<s>
+    let options = item.option.split(" ")[0].split("|");
+    return [
+      ...a,
+      ...options.filter(
+        option => option.startsWith("--") || option.startsWith("-")
+      )
+    ];
+  }, []);
+
+  let invalidOptions = inputOptions.filter(opt => !validOptions.includes(opt));
+
+  // TODO: Remove exception for 'truffle run' when plugin options support added.
+  if (invalidOptions.length > 0 && command.name !== "run") {
+    if (options.logger) {
+      const log = options.logger.log || options.logger.debug;
+      log(
+        "> Warning: possible unsupported (undocumented in help) command line option(s): " +
+          invalidOptions
+      );
+    }
+  }
+
+  return {
+    ...clone,
+    ...command.options
+  };
+};
+
+const runCommand = async function (command, options) {
```

Comment:
Is `command` here the same thing that is returned by `getCommand`?

Reply:
Yes it is.